### DISCUSSION TITLE
ChatGPT integration (with audio output)

### DISCUSSION THREAD

**Discussion by danic85** [makerforgetech/modular-biped#18] [Ideas]
There has been a discussion recently around integrating ChatGPT in the following way:

1. Existing speech recognition publishes text in `speech` event.
2. New ChatGPT module subscribes to `speech` event and publishes text to ChatGPT via API.
3. Response from ChatGPT is published in `tts` event (output).
4. TTS module converts TTS and outputs over audio.

There is currently a limitation here as the audio output has been disabled to facilitate [Neopixels](https://github.com/dmt-labs/modular-biped/blob/v3/README.md#neopixel). 
If that were resolved, then the `v3` branch already includes a text-to-speech (TTS) module that is configured to output anything published to the topic `tts`:

`pub.sendMessage('tts', msg='I am awake')`

Therefore the new ChatGPT model could subscribe to `speech` and publish `tts`. No further integration would be required.

Reference:
[Creating a module](https://github.com/dmt-labs/modular-biped/wiki/Software:--Creating-a-module)

**Comment by WhoIsMrSentry** [comment]
A little footnote from what I researched with Dan.As you will use it as sound output do not use raspberry pi. 
Raspberry pi neopixel leds incompatible with it you need to turn off the sound for it to work properly use cards like jetson nano.

**Comment by danic85** [comment]
Thanks @SentryCoderDev . I have created another idea to discuss the audio output on the Pi:

https://github.com/dmt-labs/modular-biped/discussions/19

We can continue discussing the ChatGPT use case here.

Feel free to create a PR if you create the new module as it would be a great addition. Don't include any private keys or access tokens in the repository, obviously.

**Reply by WhoIsMrSentry** [reply]
Hi, after months of research, I found a chatgpt site that can be used without a free API and wrote sample code and created a new PR.

[#55 ](https://github.com/dmt-labs/modular-biped/pull/55)

**Reply by danic85** [reply]
Nice! I'll take a look when I'm back at a computer 👏

**Comment by danic85** [comment]
Integration merged to develop https://github.com/makerforgetech/modular-biped/blob/develop/docs/ChatGPT.md
